apollo-federation/src/sources/connect/validation/connect/http.rs

```diff
@@ -396,18 +400,30 @@ fn validate_absolute_connect_url(
             .collect(),
     };
 
-    let Some(Part::Constant(first)) = url.parts.first() else {
+    let Part::Constant(first) = url.parts.iter().take_while(|part| matches!(part, Part::Constant(_)))
+        .fold(Part::Constant(Constant::default()),|acc, part| {
+            match (acc, part) {
+                (Part::Constant(acc), Part::Constant(extra)) => {
+                    Part::Constant(Constant { 
+                        value: acc.value + &extra.value, 
+                        location: acc.location.start..extra.location.end 
+                    })
+                },
+                _ => unreachable!("All parts are constant for sure")
+            }
+        }) else {
         return Err(relative_url_error());
     };
```

Comment:
```suggestion
    let before_first_expression: String = url
        .parts
        .iter()
        .map_while(|part| {
            if let Part::Constant(constant) = part {
                Some(&constant.value)
            } else {
                None
            }
        })
        .join("");
    if before_first_expression.is_empty() {
        return Err(relative_url_error());
    }
```

I don't think we actually need to accumulate a `Part`, so we could do something like this which I think both simplifies and gets rid of the panic

Reply:
(Could switch to `map_while` and keep `fold` if we do actually need the locations later)

Reply:
It gets rid of the `unreachable!` which I view as a landmine. I'd always rather use the type system to represent what the data is doing so we never get an accidental panic in the future

Reply:
It doesn't really:
`map_while` replaces `take_while`, that only takes when `Part::Constant(_)`, but the `unreachable` is in the fold, which is needed for the locations. I could get rig of the unreachable with the following (BUT IT TAKES A CLONE):

```rust
let first_range = url.parts.first().map(|part| part.location().start).unwrap_or_default();
    let Part::Constant(first) = Part::Constant(url
        .parts
        .iter()
        .map_while(|part| match part {
            Part::Constant(constant) => Some((constant.value.clone(), constant.location.clone())),
            _ => None
        })
        .fold((String::new(), first_range..first_range), |acc, part| {
            (acc.0 + &part.0, acc.1.start..part.1.end)
        }).into())
```

Reply:
The `constant` in `Part::Constant` is a struct—so you could return that instead of the tuple? _If_ we need locations... I'm not sure we do?